File: sdk-java-blocking/src/test/java/dev/restate/sdk/blocking/GetStateTest.java
package dev.restate.sdk.blocking;

import dev.restate.sdk.core.StateKey;
import dev.restate.sdk.core.TypeTag;
import dev.restate.sdk.core.impl.GetStateTestSuite;
import dev.restate.sdk.core.impl.testservices.GreeterGrpc;
import dev.restate.sdk.core.impl.testservices.GreetingRequest;
import dev.restate.sdk.core.impl.testservices.GreetingResponse;
import io.grpc.BindableService;
import io.grpc.stub.StreamObserver;

class GetStateTest extends GetStateTestSuite {

  private static class GetStateGreeter extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {
    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      String state =
          restateContext().get(StateKey.of("STATE", TypeTag.STRING_UTF8)).orElse("Unknown");

      responseObserver.onNext(GreetingResponse.newBuilder().setMessage("Hello " + state).build());
      responseObserver.onCompleted();
    }
  }

  @Override
  protected BindableService getStateGreeter() {
    return new GetStateGreeter();
  }
}

File: sdk-kotlin/src/test/kotlin/dev/restate/sdk/kotlin/GetStateTest.kt
package dev.restate.sdk.kotlin

import dev.restate.sdk.core.StateKey
import dev.restate.sdk.core.TypeTag
import dev.restate.sdk.core.impl.GetStateTestSuite
import dev.restate.sdk.core.impl.testservices.GreeterGrpcKt
import dev.restate.sdk.core.impl.testservices.GreetingRequest
import dev.restate.sdk.core.impl.testservices.GreetingResponse
import dev.restate.sdk.core.impl.testservices.greetingResponse
import io.grpc.BindableService
import kotlinx.coroutines.Dispatchers

internal class GetStateTest : GetStateTestSuite() {
  private class GetStateGreeter :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val state: String =
          restateContext().get(StateKey.of("STATE", TypeTag.STRING_UTF8)) ?: "Unknown"
      return greetingResponse { message = "Hello $state" }
    }
  }

  override fun getStateGreeter(): BindableService {
    return GetStateGreeter()
  }
}
